tests/jballerina-unit-test/src/test/resources/test-src/types/uniontypes/union_types_basic.bal

```diff
@@ -325,6 +325,63 @@ function testUnionWithDecimalFiniteTypes() {
     assertEquality(4.3d, <decimal>value + 2.3d);
 }
 
+type UnionTypeWithArray1 int|UnionTypeWithArray1[];
+type intArray int[];
+type UnionTypeWithArray3 UnionTypeWithArray1|string|intArray;
+
+function testRecursiveUnionTypeWithArray() {
+    UnionTypeWithArray1 a = [1, 2, 3];
+    assertEquality(true, a == [1, 2, 3]);
+
+    UnionTypeWithArray3 b = [[1, 2, 3]];
+    assertEquality(true, b == [[1, 2, 3]]);
+}
+
+type UnionTypeWithMap1 int|map<UnionTypeWithMap1>;
+type UnionTypeWithMap2 int|map<UnionTypeWithMap2>|UnionTypeWithMap1;
+
+function testRecuriveUnionTypeWithMap() {
+    UnionTypeWithMap1 a = {"one":1, "two":2};
+    assertEquality(true, a == {"one":1, "two":2});
+    UnionTypeWithMap2 b =  {"one":{"one":1, "two":2}};
+    assertEquality(true, b == {"one":{"one":1, "two":2}});
+}
+
+type UnionTypeWithMapAndArray int|UnionTypeWithMapAndArray[]|map<UnionTypeWithMapAndArray>;
+
+function testRecursiveUnionTypeWithArrayAndMap() {
+    UnionTypeWithMapAndArray a = {"one":1, "two":2};
+    assertEquality(true, a == {"one":1, "two":2});
+
+    UnionTypeWithMapAndArray b = [1, 2, 3];
+    assertEquality(true, b == [1, 2, 3]);
+
+    UnionTypeWithMapAndArray c = {"one" : [1,2,3]};
+    assertEquality(true, c == {"one" : [1,2,3]});
+
+    UnionTypeWithMapAndArray d = [{"one":1, "two":2}];
+    assertEquality(true, d ==[{"one":1, "two":2}]);
+}
+
+type UnionTypeWithRecord int|record {UnionTypeWithRecord u;};
+
+function testRecursiveUnionTypeWithRecord() {
+    UnionTypeWithRecord a = {u: 5};
+    assertEquality(true, a == {u: 5});
+    UnionTypeWithRecord b = {u: {u:5}};
+    assertEquality(true, b == {u: {u:5}});
+}
+
+type UnionTypeWithTuple1 int|[UnionTypeWithTuple1];
+type UnionTypeWithTuple2 int|[UnionTypeWithTuple2, UnionTypeWithTuple1];
```

Comment:
Shall we also add tests with recursive type as a rest type in a tuple? Can also test similar cases in records with rest type.